runtime/lib/backend/openqasm/OpenQasmBuilder.hpp

```diff
@@ -0,0 +1,409 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <iomanip>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "Exception.hpp"
+
+namespace Catalyst::Runtime::Device::OpenQasm {
+
+enum class VariableType : uint8_t {
+    Float, // = 0
+};
+
+enum class RegisterMode : uint8_t {
+    Alloc, // = 0
+    Slice,
+    Reset,
+};
+
+enum class RegisterType : uint8_t {
+    Qubit, // = 0
+    Bit,
+};
+
+using GateNameT = std::string_view;
+
+/**
+ * The map of supported quantum gate names by OpenQasm devices.
+ */
+constexpr std::array rt_qasm_gate_map = {
+    // (RT-GateName, Qasm-GateName)
+    std::tuple<GateNameT, GateNameT>{"PauliX", "x"},
+    std::tuple<GateNameT, GateNameT>{"PauliY", "y"},
+    std::tuple<GateNameT, GateNameT>{"PauliZ", "z"},
+    std::tuple<GateNameT, GateNameT>{"Hadamard", "h"},
+    std::tuple<GateNameT, GateNameT>{"S", "s"},
+    std::tuple<GateNameT, GateNameT>{"T", "t"},
+    std::tuple<GateNameT, GateNameT>{"CNOT", "cnot"},
+    std::tuple<GateNameT, GateNameT>{"CZ", "cz"},
+    std::tuple<GateNameT, GateNameT>{"SWAP", "swap"},
+    std::tuple<GateNameT, GateNameT>{"PhaseShift", "phaseshift"},
+    std::tuple<GateNameT, GateNameT>{"RX", "rx"},
+    std::tuple<GateNameT, GateNameT>{"RY", "ry"},
+    std::tuple<GateNameT, GateNameT>{"RZ", "rz"},
+    std::tuple<GateNameT, GateNameT>{"CSWAP", "cswap"},
+    std::tuple<GateNameT, GateNameT>{"PSWAP", "pswap"},
+    std::tuple<GateNameT, GateNameT>{"ISWAP", "iswap"},
+    std::tuple<GateNameT, GateNameT>{"Toffoli", "ccnot"},
+};
+
+/**
+ * Lookup OpenQasm gate names.
+ */
+constexpr auto lookup_qasm_gate_name(std::string_view gate_name) -> std::string_view
+{
+    for (auto &&[gate_qir, gate_qasm] : rt_qasm_gate_map) {
```

Comment:
I didn't know this kind of matching is possible, must be something new, thanks!

Reply:
Structured bindings :) https://en.cppreference.com/w/cpp/language/structured_binding